### ISSUE TITLE
Wavetable "morph" smoothing lag

### ISSUE THREAD

**Issue by blancoberg** [surge-synthesizer/surge#7696]
**Bug Description:**
The wavetable morph parameter has a ~15ms smoothing lag.
Compare to other parameters like pitch, the morph parameter lerps to the new value instead of setting in somewhat directly. 

**Surge XT Version**
* Version: 1.3.main.f4c97fc
* Plugin Type: All
* Bitness: 64-bit

**Reproduction Steps:**
* Select the wavetable oscillator
* Switch to Mono ST mode 
*  Connect the note modulator to the "Morph" parameter at 100%. 
* Play legato notes with c4 and c5 in order to switch between the first and last frame of the wavetable. It should be audible that the wavetable frame position does not switch instantly. It drags behind the note change. 
* If you can't hear it, try to disable keytracking and just listen to the glide between the frames. 

Removing this line in "WaveTableOscillator.cpp":
``` 
float shape = l_shape.v;
``` 
And replacing it with this:
``` 
float shape = limit_range(localcopy[id_shape].f, 0.f, 1.f); 
``` 
Fixes the issue. ( This was suggested by @baconpaul  for testing )

**Expected Behavior:**
The morph parameter should change at the same speed as any other parameter, with as little lag as possible. 

**Computer Information (please complete the following!):**
 - OS: Windows 10
 - Host: Reaper
 - Version: 

**Additional Information:**

**Comment by blancoberg** [comment]
this test is a wavetable that plays a sine wave at 2x speed until it arrives to the last frame where it runs at 1x with any interpolation in between the frames.


How to use the test preset:
When playing a legato note between C4 and C5, it should stay on C4. The note should not have time to switch before arriving to the last frame. 

Demonstration:
https://youtu.be/FxMLi7NjKuc

[Morph lag test.zip](https://github.com/user-attachments/files/16158362/Morph.lag.test.zip)

**Comment by blancoberg** [comment]
I don't know if this issue is related, but I suspect it is. 

Other than having a smoothing lag, there is an inconsistency when modulating the "morph" parameter. If you for example modulate the morph parameter with the modwheel, Its expected that the morph value should start on the value that the modwheel has on note start. 

However, sometimes it seems like it take a whole block for it to set the value, causing a glitch because its essentially scrubbing through the entire wavetable. 

expected result:
![image](https://github.com/user-attachments/assets/53a789d5-ac0e-4171-bc90-bb72eb4fe737)

But it sometimes produces this: ( it looks like it is about 30 samples long)
![image](https://github.com/user-attachments/assets/86847a02-0bac-44fd-914e-92c3e8a3ff3a)

Here is a patch for testing: 
[wavetable morph glitch.zip](https://github.com/user-attachments/files/16475823/wavetable.morph.glitch.zip)

**Comment by baconpaul** [comment]
Oh that may be something different I bet. I suspect the mod wheel smoother starts at zero not at last-value in the midi handler when a new voice triggers. We should instantize it at voice on, but then we need to know the 'last' mod wheel, which I think we have somewhere, but I'm not 100% sure.

**Comment by blancoberg** [comment]
I don't think it has anything to do with the modwheel, that was just one example. 
I'v have tried it with multiple modulators and they all seem to cause it. And it it just the morph parameter as far as I can tell.

**Comment by blancoberg** [comment]
its seems like the morph parameter doesnt set an initial value, so it just waits until the next block. 

If you look at the waveform, the glitch looks like it "scrubs" through the wavetable during one block period.

**Comment by baconpaul** [comment]
Gotcha. And yeah that also makes sense

I tried to fix this but it had some other unintended consequences. Basically replacing all the smooths with just lerps is sound-changing. So still thinking about this one. But its definitely wrong.

**Comment by blancoberg** [comment]
I don't think it can be changed without affecting the sound to some extend. 

Only safe way to change it would be to add a setting for the patch were you can turn off wavetable smoothing manually. its on by default.

**Comment by baconpaul** [comment]
Yes that’s what we plan to do but it’s a bit tricky

**Comment by blancoberg** [comment]
Maybe a lag setting could be added in general for all modulation sources? Kind of like in kontakt has.
All modulation sources does have a setLag function, I'm thinking it would be awesome to be able to adjust that setting from 0% - 100% ( which translates to 1 to 0.01 or something), and also have a way of reverting back to the initial value that particular modulation has.

Just right click, click adjust lag, the same way as adjusting any other parameter. Only thing missing there UI wise is some kind of button for resetting the value to its initial state. 
![image](https://github.com/user-attachments/assets/49cd5095-b3cc-4466-b59a-792c234a7ec3)

**Comment by mkruselj** [comment]
The general plan is to have modulation mappers in XT 2.0, which will also allow you to do lag (and obviously reshape the modulation with a custom curve, think like Zebra's modmappers).

**Comment by blancoberg** [comment]
But doesnt all modulation sources already have lag built in? so it shouldnt add any more cpu than it already is. Atleast all oscillators have setLag() functions on all its sliders.

I'm pretty sure this issue is mostly about that lag, since it is pretty much fixed when you adjust the lag in code

**Comment by blancoberg** [comment]
I'm referring to these settings that most oscillators has:
```
float rate = 0.05;
l_shape.setRate(rate);
l_clip.setRate(rate);
l_vskew.setRate(rate);
l_hskew.setRate(rate);
```

**Comment by baconpaul** [comment]
some use lag some use lerp across the block

the change i wanted to make to wavetable was to have an option to lerp rather than lag and then it really is just across-the-block instant application.

that change is a bit tricky though.

**Comment by mkruselj** [comment]
> I'm referring to these settings that most oscillators has:
> 
> ```
> float rate = 0.05;
> l_shape.setRate(rate);
> l_clip.setRate(rate);
> l_vskew.setRate(rate);
> l_hskew.setRate(rate);
> ```

No. This is smoothing the changes of modulator parameter changess, not smoothing the output of the modulator. Definitely not the same thing.

**Comment by baconpaul** [comment]
> > I'm referring to these settings that most oscillators has:
> > ```
> > float rate = 0.05;
> > l_shape.setRate(rate);
> > l_clip.setRate(rate);
> > l_vskew.setRate(rate);
> > l_hskew.setRate(rate);
> > ```
> 
> No. This is smoothing the changes of oscillator parameter changes, not smoothing the output of the modulator. Definitely not the same thing.

well defacto it is the same thing. since those changes include the modulation source and it is what causes the problem.

i agree we don't want to smooth the modulator output, but control over that smoothing is the core of the issue here. it's too slow especially now the other issue with table lookup is getting resolved.

**Comment by blancoberg** [comment]
> i agree we don't want to smooth the modulator output, but control over that smoothing is the core of the issue here. it's too slow especially now the other issue with table lookup is getting resolved.

Exactly this. With the new morph interpolation, I honestly can't hear any difference in sound when modulating morph with lfos, even with that lag setting set to zero. its really not needed ( with the new interpolation that is )

The only issue when setting the lag to zero is that modulations like modwheel, aftertouch or any other values that is changed with user input, will get some stuttering. So in cases were you do set the lag to 0, those inputs needs to be smoothed out instead, which can already be done with a lua modulator for instance

**Comment by mkruselj** [comment]
It includes the _sum_ of all modulations targetting that parameter.

I don't think this param smoothing should be granularly adjustable per parameter. That's too much and too esoteric. Find a good value for the affected parameter (in this case Morph) and have an if for it, that's it.

**Comment by baconpaul** [comment]
The midi inputs are already smoothed by the way. I think it is just direct learn cc which would be a problem

**Comment by baconpaul** [comment]
But I agree setting the rate to something just a smidge above 0 and sticking it in the if is the way to go

**Comment by blancoberg** [comment]
I think we should set it to zero, as there is no audible difference when modulating with lfos and such. 
With midi input with smoothing set to exp or no smoothing, there is some noise but only when modulating the worse case scenario wavetable that phase shifts like crazy. 

In fast linear smooth mode, it works like a charm.

**Comment by mkruselj** [comment]
There should absolutely be _some_ smoothing applied, as this would not impact only MIDI input but also host automation and UI gestures.

**Comment by baconpaul** [comment]
Most other oscillators lerp over the block

that was what I wanted to do here but I couldn’t figure it out

I agree zero is probably the wrong setting but I think the bigger thing is lag vs lerp smoothing is the problem

**Comment by baconpaul** [comment]
The thing is we really don’t want to smooth inside the dsp since we don’t know if the source is smooth needed or not.

this is a pretty messy general problem with surge though

**Comment by blancoberg** [comment]
it does lerp now during the block though? its not enough though as the ui interaction and daw automation is way lower in frequency. 

For this to work, the lag has to not affect the entire parameter, only the user interaction. E.g when you interact with the slider by hand or with daw automation. Regular modulations connected to it shouldnt have any smoothing as it is not needed, and midi smoothing solves the rest. 

Basically the handle of the slider needs smoothing. not the slider value itself, if that makes sense

**Comment by blancoberg** [comment]
Maybe an exception could be implemented in the parameter, like onlyLagUserInput = true and then implement logic that only adds the lag to the "handle" of the slider but all other modulation values are just retreived directly

**Comment by baconpaul** [comment]
So some inputs - notably midi controller and macro inputs - are already smoothed and that smoothed value is updated every block.  You still want to lerp across the block in many cases (like you don' want a filter cutoff to have a 32 block step size).

Others like user interactions in the gui and vst3 parameter sets are not

this is a problem in surge more broadly than here. Some parts of the dsp internally smooth rather than fix that problem. This gives the problems you have found.

There is a good solution but it rhymes with "re-write". Because it is a re-write of the parts which are a problem.

For this situation I'm not sure what the right answer is. My gut is 'a very small smooth in an if and fix the other bug in the same if once i get a linux booted to test it' is right. but that's not a great answer either since the correct answer isn't available in this iteration of the codebase.

**Comment by mkruselj** [comment]
Done now, so closing.
